File: Sorting/bubble_sort.cpp
#include<iostream>

//functions used
void bubble_sort(int *a, int n);
void modified_bubble_sort(int *a, int n);

//main method
int main(){
    int N;
    std::cin>>N;
    int A[N];
    for(int i=0; i<N; i++){
        std::cin>>A[i];
    }
    modified_bubble_sort(A,N);
    //bubble_sort(A,N);
    for(int i=0; i<N; i++){
        std::cout<<A[i]<<" ";
    }
    std::cout<<std::endl;
    return 0;
}

//normal bubble sort
void bubble_sort(int *a, int n){
    for(int i=0; i<n; i++){
        for(int j=0; j<n-1; j++){
            if(a[j]>a[j+1]){
                //swap
                int temp;
                temp=a[j];
                a[j]=a[j+1];
                a[j+1]=temp;
            }
        }
    }
}

//modified bubble sort
void modified_bubble_sort(int *a, int n){
    for(int i=0; i<n-1; i++){
    	int flag=0;
        for(int j=0; j<n-1; j++){
            if(a[j]>a[j+1]){
                //swap
                int temp;
                temp=a[j];
                a[j]=a[j+1];
                a[j+1]=temp;
                flag=1;
            }
        }
        if(flag==0) break;
    }
}
/*
NOTE: in modified bubble sort, if the array get sorted after k'th pass then in (k+1)'th pass the if condition become true and the outer for loop will stop from unnecessary checking process
*/

File: Data Structure/Sorting/Bubble Sort/bubbleSort.py
def bubbleSort(A):
    ''' Takes an array A and sort inplace
    Args:   A (list) = array of number
    Return: None
    '''
    l = len(A)
    while l >= 0:                               # outer loop
        flag = False
        for i in range(l-1):                    # inner loop
            if A[i] > A[i+1]:
                A[i], A[i+1] = A[i+1], A[i]     # swap
                flag = True
        l -= 1
        if flag == False:                       # for sorted list
            break

# >>> import bubbleSort
# >>> a = [1,2,6,5,4,3]
# >>> bubbleSort.bubbleSort(a)
# >>> a
# [1, 2, 3, 4, 5, 6]
